src/java/org/apache/cassandra/cql3/functions/CollectionFcts.java

```diff
@@ -0,0 +1,364 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.cql3.functions;
+
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import com.google.common.collect.ImmutableList;
+
+import org.apache.cassandra.cql3.CQL3Type;
+import org.apache.cassandra.db.marshal.AbstractType;
+import org.apache.cassandra.db.marshal.CollectionType;
+import org.apache.cassandra.db.marshal.Int32Type;
+import org.apache.cassandra.db.marshal.ListType;
+import org.apache.cassandra.db.marshal.MapType;
+import org.apache.cassandra.db.marshal.SetType;
+import org.apache.cassandra.transport.ProtocolVersion;
+
+/**
+ * Native CQL functions for collections (sets, list and maps).
+ * <p>
+ * All the functions provided here are {@link NativeScalarFunction}, and they are meant to be applied to single
+ * collection values to perform some kind of aggregation with the elements of the collection argument. When possible,
+ * the implementation of these aggregation functions is based on the accross-rows aggregation functions available on
+ * {@link AggregateFcts}, so both across-rows and within-collection aggregations have the same behaviour.
+ */
+public class CollectionFcts
+{
+    public static void addFunctionsTo(NativeFunctions functions)
+    {
+        functions.add(new FunctionFactory("map_keys", FunctionParameter.anyMap())
+        {
+            @Override
+            protected NativeFunction doGetOrCreateFunction(List<AbstractType<?>> argTypes, AbstractType<?> receiverType)
+            {
+                return makeMapKeysFunction(name().name, (MapType<?, ?>) argTypes.get(0));
+            }
+        });
+
+        functions.add(new FunctionFactory("map_values", FunctionParameter.anyMap())
+        {
+            @Override
+            protected NativeFunction doGetOrCreateFunction(List<AbstractType<?>> argTypes, AbstractType<?> receiverType)
+            {
+                return makeMapValuesFunction(name().name, (MapType<?, ?>) argTypes.get(0));
+            }
+        });
+
+        functions.add(new FunctionFactory("collection_count", FunctionParameter.anyCollection())
+        {
+            @Override
+            protected NativeFunction doGetOrCreateFunction(List<AbstractType<?>> argTypes, AbstractType<?> receiverType)
+            {
+                return makeCollectionCountFunction(name().name, (CollectionType<?>) argTypes.get(0));
+            }
+        });
+
+        functions.add(new FunctionFactory("collection_min", FunctionParameter.setOrList())
+        {
+            @Override
+            protected NativeFunction doGetOrCreateFunction(List<AbstractType<?>> argTypes, AbstractType<?> receiverType)
+            {
+                return makeCollectionMinFunction(name.name, (CollectionType<?>) argTypes.get(0));
```

Comment:
Could we unify uses of `name()` and `name`?

Reply:
We're not doing this one?

Reply:
Yep, I forgot about this bit. Done.